### PR TITLE
DM-15024: Update package and module documentation directory roles

### PR DESCRIPTION
**Pull request by jonathansick**
This updates the design to reflect our new thinking on the roles of package and module documentation directories (package doc dirs are only used for data-only packages; otherwise documentation content resides in module documentation directories.

This PR also adds a notice indicating that the DM Developer Guide has priority for implementation documentation and that readers shouldn't expect this document to track the implementation in the future.

https://dmtn-030.lsst.io/v/DM-15024

### CODE REVIEW

File: _docs-as-code.rst [hunk 1]
```diff
@@ -38,15 +38,16 @@ Organization of documentation in packages
...
 
 1. **Documentation for Python modules,** such as ``lsst.afw.table``.
    This documentation content is oriented towards the code base, APIs, and its usage.
    The bulk of Science Pipelines documentation is organized around Python modules.
    We describe this content further in :ref:`module-type`.
-2. **Documentation for the package itself** such as ``afw``.
-   This documentation concerns the Git repository itself and may provide developer documentation that is not oriented around Python modules.
-   In fact, some packages, like ``afwdata`` or ``verify_metrics``, do not have any APIs and are entirely documented through this type of documentation.
+
+2. **Documentation for the package** such as ``afw``.
```
Comment (ktlim) [inline _docs-as-code.rst]: Maybe use a different example, such as those below?

File: _docs-as-code.rst [hunk 2]
```diff
@@ -105,6 +100,22 @@ These could be PDFs or tarball downloads.
 This static content is stored in a ``_static/`` directory.
 So that static content from all packages can be integrated, each package must store static content in a sub-directory of the ``_static`` directory, such as ``_static/afw``.
 
+Data-only packages, like ``afwdata`` have package documentation directories named after the package/Git repository itself.
```
Comment (ktlim) [inline _docs-as-code.rst]: I'm not sure these are always "data-only".  We could conceivably have a "C++-only" package, for example.  Comma after `afwdata`.

File: _docs-as-code.rst [hunk 3]
```diff
@@ -129,7 +140,7 @@ When pipelines_lsst_io_ is built, the package, module, and ``_static`` documenta
         lsst.afw.cameraGeom/ -> link to /afw/doc/lsst.afw.cameraGeom/
         ..
       packages/
-        afw/ -> link to /afw/doc/afw/
+        afwdata/ -> link to /afw/doc/afwdata/
```
Comment (SimonKrughoff) [inline _docs-as-code.rst]: Is this supposed to be `/afwdata/doc/afwdata/`?

File: _docs-as-code.rst [hunk 4]
```diff
@@ -144,7 +155,8 @@ As an example, the ``doc/manifest.yaml`` file included in ``afw`` may look like
 .. code-block:: yaml
 
    # Name of the package and also name of the package doc directory
-   package: "afw"
+   # Note: N/A for `afw`.
+   # package: "afw"
```
Comment (SimonKrughoff) [inline _docs-as-code.rst]: It feels a little strange to have a commented line in the doc that can never be uncommented.  Maybe don't included a note about a package section if one cannot exist?

**Review by SimonKrughoff — APPROVED** [review]
LGTM.  Two small comments.

**Pull request merged**